File: app/src/main/java/io/github/micopiira/mylauncher/AppsAdapter.java
package io.github.micopiira.mylauncher;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

class AppsAdapter extends ArrayAdapter<AppDetail> {

    AppsAdapter(Context context, List<AppDetail> apps) {
        super(context, 0, apps);
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

        AppDetail app = getItem(position);

        ViewHolder holder;

        if (convertView != null) {
            holder = (ViewHolder) convertView.getTag();
        } else {
            convertView = LayoutInflater.from(getContext()).inflate(R.layout.list_item, parent, false);
            holder = new ViewHolder();
            holder.appLabel = (TextView) convertView.findViewById(R.id.item_app_label);
            holder.appIcon = (ImageView) convertView.findViewById(R.id.item_app_icon);

            convertView.setTag(holder);
        }

        holder.appIcon.setImageDrawable(app.getIcon());
        holder.appLabel.setText(app.getLabel());

        SharedPreferences SP = PreferenceManager.getDefaultSharedPreferences(getContext());

        holder.appLabel.setTextColor(Color.parseColor(SP.getString("label_color", "#ffffff")));

        return convertView;
    }

    private static class ViewHolder {
        ImageView appIcon;
        TextView appLabel;
    }

}

File: app/src/main/java/io/github/micopiira/mylauncher/AppsAdapter.kt
package io.github.micopiira.mylauncher

import android.content.Context
import android.graphics.Color
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*

class AppsAdapter(context: Context, private val apps: List<AppDetail>) : ArrayAdapter<AppDetail>(context, 0, apps), Filterable {

    var filteredApps = apps;

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(p0: CharSequence?): FilterResults {
                val filterResults = FilterResults()
                filterResults.values = apps.filter {
                    if (p0 != null && p0.isNotEmpty()) it.label.contains(p0, true) || it.label.contains(p0, true) else true
                }
                return filterResults;
            }

            override fun publishResults(p0: CharSequence, p1: FilterResults) {
                filteredApps = p1.values as List<AppDetail>
                notifyDataSetChanged()
            }
        }
    }


    override fun getCount(): Int {
        return filteredApps.size
    }

    override fun getItem(position: Int): AppDetail {
        return filteredApps[position]
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        val app = getItem(position)
        val holder: ViewHolder

        if (convertView != null) {
            holder = convertView.tag as ViewHolder
        } else {
            convertView = LayoutInflater.from(context).inflate(R.layout.list_item, parent, false)
            holder = ViewHolder(
                    appLabel = convertView!!.findViewById(R.id.item_app_label) as TextView,
                    appIcon = convertView.findViewById(R.id.item_app_icon) as ImageView
            )
            convertView.tag = holder
        }

        holder.appIcon!!.setImageDrawable(app.icon)
        holder.appLabel!!.text = app.label

        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

        holder.appLabel.setTextColor(Color.parseColor(sharedPreferences.getString("label_color", "#ffffff")))

        return convertView
    }

    private data class ViewHolder(
            internal val appIcon: ImageView? = null,
            internal val appLabel: TextView? = null
    )

}
